coreneuron/coreneuron.hpp

```diff
@@ -0,0 +1,238 @@
+
+using DependencyTable = std::vector<std::vector<int>>;
+
+/**
+ * A class representing the CoreNEURON state, holding pointers to the various data structures
+ *
+ * The pointers to "global" data such as the NrnThread, Memb_list and Memb_func data structures
+ * are managed here. they logically share their lifetime and runtime scope with instances of
+ * this class.
+ */
+class CoreNeuron {
+
+
+    /// Local to coreneuron, used to keep track of point process IDs
+    int pointtype = 1; /* starts at 1 since 0 means not point in pnt_map*/
+    /**
+     * map if mech is a point process
+     * In the future only a field of Mechanism class
+     */
+    std::vector<char> pnt_map; /* so prop_free can know its a point mech*/
+
+    /** Vector mapping the types (IDs) of different mechanisms of mod files between Neuron and
+     * CoreNeuron
+     */
+    std::vector<int> different_mechanism_type;
+
+    /**
+     * dependency helper filled by calls to hoc_register_dparam_semantics
+     * used when nrn_mech_depend is called
+     * vector-of-vector DS. First idx is the mech, second idx is the dependent mech.
+     */
+    DependencyTable ion_write_dependency;
+
+    std::vector<Memb_func> memb_funcs;
+
+    /**
+     * Net send / Net receive
+     * only used in crnrn for book keeping synapse mechs, should go into Coreneuron class
+     */
+    std::vector<std::pair<NetBufReceive_t, int>> net_buf_receive;
+    std::vector<int> net_buf_send_type;
+
+    /**
+     * before-after-blocks from nmodl are registered here as function pointers
+     */
+    std::array<BAMech*, BEFORE_AFTER_SIZE> bamech;
+
+    /**
+     * Internal lookup tables. Number of float and int variables in each mechanism and memory layout
+     * future --> mech class
+     */
+    std::vector<int> nrn_prop_param_size_;
+    std::vector<int> nrn_prop_dparam_size_;
+    std::vector<int> nrn_mech_data_layout_; /* 1 AoS (default), >1 AoSoA, 0 SoA */
+    /* array is parallel to memb_func. All are 0 except 1 for ARTIFICIAL_CELL */
+    std::vector<short> nrn_artcell_qindex_;
+    std::vector<bool> nrn_is_artificial_;
+
+    /**
+     * Net Receive function pointer lookup tables
+     */
+    std::vector<pnt_receive_t> pnt_receive; /* for synaptic events. */
+    std::vector<pnt_receive_t> pnt_receive_init;
+    std::vector<short> pnt_receive_size;
+
+    /**
+     * Holds function pointers for WATCH callback
+     */
+    std::vector<nrn_watch_check_t> nrn_watch_check;
+    /**
+     * values are type numbers of mechanisms which do net_send call
+     * related to NMODL net_event()
+     *
+     */
+    std::vector<int> nrn_has_net_event_;
+    /**
+     * inverse of nrn_has_net_event_ maps the values of nrn_has_net_event_ to the index of
+     * ptntype2presyn
+     */
+    std::vector<int> pnttype2presyn;
+
+
+    std::vector<bbcore_read_t> nrn_bbcore_read_;
+    std::vector<bbcore_write_t> nrn_bbcore_write_;
+
+  public:
+
+    auto& get_memb_funcs() {
```

Comment:
All very good points! I am only allowing these rather ugly accessors because this PR is in fact an intermediate step. Many of those accessors will disappear in the near future because the data will be accessed through class methods of the owning class or new, proper interfaces will be implemented.
I wanted to just move the data as quickly as possible into a class and start encapsulate them.
If it's OK, I would defer the changes you suggest one of the next PRs.